src/main/kotlin/com/sourcegraph/cody/chat/ui/ChatPanel.kt

```diff
@@ -94,6 +113,24 @@ class ChatPanel(project: Project, chatSession: ChatSession, modelFromState: Chat
 
   fun updateWithSessionId(sessionId: SessionId) {
     llmDropdown.fetchAndUpdateModels(sessionId)
-    llmDropdown.sessionId.complete(sessionId)
+  }
+
+  private fun setLlmForAgentSession(chatModelProvider: ChatModelsResponse.ChatModelProvider) {
+    val sessionId = chatSession.getSessionId() ?: return
+
+    CodyAgentService.withAgentRestartIfNeeded(project) { agent ->
+      val activeAccountType = CodyAuthenticationManager.instance.getActiveAccount(project)
+      if (activeAccountType.isEnterpriseAccount()) {
+        agent.server.webviewReceiveMessage(
+            WebviewReceiveMessageParams(sessionId, WebviewMessage(command = "chatModel")))
+      } else {
+        agent.server.webviewReceiveMessage(
+            WebviewReceiveMessageParams(
+                sessionId, WebviewMessage(command = "chatModel", model = chatModelProvider.model)))
+      }
+    }
+
+    HistoryService.getInstance(project)
+        .updateChatLlmProvider(chatSession.getInternalId(), chatModelProvider)
```

Comment:
instead of updating it every time the messages change let's update it on selected item change only